### PR TITLE
Forbidding zombies from harming other zombies

### PR DESCRIPTION
**Pull request by John-Doom**
:cl:
- tweak: Zombies can't harm other zombies.

### CODE REVIEW

**Comment by Aisu9** [comment]
this could be extender to most consoles and ape/other machine that distribute power, chemdispenser/chemaster
they are supposed to be too dumb to know which machine is vital or not to the station

**Comment by ficcialfaint** [comment]
I think it should be put in CL so players do not wonder why they can't destroy them.

File: Content.Shared/Zombies/ZombieComponent.cs
Comment (ficcialfaint) [inline Content.Shared/Zombies/ZombieComponent.cs:22]: Why'd it be removed?

File: Content.Server/Zombies/ZombieSystem.cs
Comment (ficcialfaint) [inline Content.Server/Zombies/ZombieSystem.cs:307]:
```suggestion
        private bool GetIsForbiddenTarget(EntityUid entity)
```

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

**Comment by HoidC** [comment]
> what about forbidding them from destroying any type of walls too? zombies should only go after crew members and not eating every walls of the station

breaking walls may be necessary to reach the crew and I don't think zombies ripping apart every wall is an issue rn

**Review by mirrorcult — CHANGES_REQUESTED** [review]
this is the wrong solution. if you want to make these forbidden to destroy then make them electrocute on melee attack, like APCs/substations/SMES already do

File: Content.Shared/Zombies/ZombieComponent.cs
Comment (John-Doom) [reply Content.Shared/Zombies/ZombieComponent.cs:22]: because i removed its use from ZombieSystem.cs

Comment (ficcialfaint) [reply Content.Shared/Zombies/ZombieComponent.cs:22]: Eh? Your PR says it's only about making zombies not able to destroy AME and Gravity

**Review by ficcialfaint — CHANGES_REQUESTED** [review]
Do NOT webedit
image

**Comment by John-Doom** [comment]
> also _even if_ this was the right way to go about it, having zombie reference ame directly is very very very very bad use an event

Which events do you refer to? Sorry, I'm not familiar with the engine yet, I basically started 5 hours ago xd

**Comment by Myakot** [comment]
Gravity gen is already immortal.

File: Content.Shared/Zombies/ZombieComponent.cs
Comment (VasilisThePikachu) [reply Content.Shared/Zombies/ZombieComponent.cs:22]: Why did you remove it exactly? is it interfering with anything on your code? You just dont like it?

**Comment by malchanceux** [comment]
Remember the shitter when making changes like this.
image

**Comment by deltanedas** [comment]
i died pls restart

File: Content.Shared/Zombies/ZombieComponent.cs
Comment (John-Doom) [reply Content.Shared/Zombies/ZombieComponent.cs:22]: I did replace it: the original source let a zombie still make damage do another zombie using that coefficent, I made it so it can't deal damage at all.

**Comment by John-Doom** [comment]
> this is the wrong solution. if you want to make these forbidden to destroy then make them electrocute on melee attack, like APCs/substations/SMES already do

It seems a bit too punishing to me, could get in the way during a legit fight. Are you sure?

**Comment by Aisu9** [comment]
not like zombies go out of their ways most of the time, spacing the station, destroying vital points and appliances of the station. i'm sure they'll start welder bombing if this merge on WizDen

**Review by John-Doom — COMMENTED** [review]
removed gravity check since it's already immortal, succesfully tested in-game

**Comment by John-Doom** [comment]
> not like zombies go out of their ways most of the time, spacing the station, destroying vital points and appliances of the station. i'm sure they'll start welder bombing if this merge on WizDen

If they do, I'll add more exceptions.

**Comment by metalgearsloth** [comment]
> > this is the wrong solution. if you want to make these forbidden to destroy then make them electrocute on melee attack, like APCs/substations/SMES already do
> 
> It seems a bit too punishing to me, could get in the way during a legit fight. Are you sure?

Yes.

**Comment by John-Doom** [comment]
> > > this is the wrong solution. if you want to make these forbidden to destroy then make them electrocute on melee attack, like APCs/substations/SMES already do
> > 
> > 
> > It seems a bit too punishing to me, could get in the way during a legit fight. Are you sure?
> 
> Yes.

Could someone point me to the class which electrocutes zombies? I'm having a hard time finding it :(

**Comment by tom-leys** [comment]
> 
> Could someone point me to the class which electrocutes zombies? I'm having a hard time finding it :(

Have a look at the change which introduced it here : https://github.com/space-wizards/space-station-14/pull/16584

**Comment by John-Doom** [comment]
> > Could someone point me to the class which electrocutes zombies? I'm having a hard time finding it :(
> 
> Have a look at the change which introduced it here : #16584

So the file was already setup to electrocute on damage, but doesn't work for some reason o.O

**Comment by John-Doom** [comment]
the bug seems to be in ElectrocutionSystem.cs at line 269:
`tryNode.NodeGroup is IBasePowerNet { NetworkNode: { LastCombinedSupply: > 0 } })`

**Comment by John-Doom** [comment]
If I let the condition pass through, the game crashes at https://github.com/space-wizards/space-station-14/blob/4783a655a48f343fe297ca6679211d3426bd1b00/Content.Server/Electrocution/ElectrocutionSystem.cs#LL322C4-L322C4

**Comment by VasilisThePikachu** [comment]
Just making sure, you are not wearing captain gloves when testing the electrocution? They are insulated

**Comment by HoidC** [comment]
> Just making sure, you are not wearing captain gloves when testing the electrocution? They are insulated
> 
> Also probably better to join the discord for this and get help in the #howdoicode channel, I'm not an expert contributor but referencing ame in electrocution system seems like the wrong way to do it

don't zombies autodrop gloves

**Comment by VasilisThePikachu** [comment]
> > Just making sure, you are not wearing captain gloves when testing the electrocution? They are insulated
> > 
> > Also probably better to join the discord for this and get help in the #howdoicode channel, I'm not an expert contributor but referencing ame in electrocution system seems like the wrong way to do it
> 
> don't zombies autodrop gloves

Even if they do, I still feel like there is a better way then referencing ame in electrocution system directly

**Comment by deltanedas** [comment]
dont apc sub and smes have components for being electrified

**Comment by John-Doom** [comment]
> dont apc sub and smes have components for being electrified

And so have AME shields, but it seems the current source doesn't make use of them, as I said: https://github.com/space-wizards/space-station-14/pull/16963#issuecomment-1571742420
No, my actual question here is why github fails to "build & test debug", while it works fine locally :\

**Comment by John-Doom** [comment]
I don't understand why it keeps failing here: https://github.com/space-wizards/space-station-14/actions/runs/5146945641/jobs/9266690219?pr=16963#step:10:16

**Comment by 0x6273** [comment]
> Prices have nothing to do with my changes, I don't understand why github keeps failing here: https://github.com/space-wizards/space-station-14/actions/runs/5146945641/jobs/9266690219?pr=16963#step:10:16

Someone merged a broken PR so now master is broken and tests are failing on all PRs.

**Comment by John-Doom** [comment]
@mirrorcult @PJB3005 I'm waiting for your approval o.O

**Comment by metalgearsloth** [comment]
Ignore the reviewers tab it's meaningless.

File: Content.Server/Electrocution/ElectrocutionSystem.cs
Comment (metalgearsloth) [inline Content.Server/Electrocution/ElectrocutionSystem.cs:270]: Nodegroups shouldn't be hardcoded in here.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Shouldn't you just make a friendlyfire faction component.

File: Content.Server/Electrocution/ElectrocutionSystem.cs
Comment (John-Doom) [reply Content.Server/Electrocution/ElectrocutionSystem.cs:270]: Check here: https://github.com/space-wizards/space-station-14/pull/16963#issuecomment-1571742420

Comment (John-Doom) [reply Content.Server/Electrocution/ElectrocutionSystem.cs:270]: I wasn't aware of such component, I'll see if I can make use of it o.O

**Comment by John-Doom** [comment]
> Shouldn't you just make a friendlyfire faction component.

Do you mean these classes? They seem to be for the AIs, not prevent friendly fire between players :\
https://github.com/space-wizards/space-station-14/blob/9852bba84776ad914f5607bb00a393c0cd0aff29/Content.Server/NPC/Components/FactionComponent.cs#L8
https://github.com/space-wizards/space-station-14/blob/821f4cc3d4427510869894ea3a97d88a1bd418f7/Content.Server/NPC/Systems/FactionSystem.cs#L10

File: Content.Server/Zombies/ZombieSystem.cs
Comment (Skyedra) [inline Content.Server/Zombies/ZombieSystem.cs:257]: Based on your other comments, I'm not sure if this is doing what you meant to do.  Previously, zombie friendly fire damage was limited by the coefficient (see how it multiplies it by .25, basically making it quarter damage)... with this change, it seems like zombie vs zombie would do full damage again.

Comment (John-Doom) [reply Content.Server/Zombies/ZombieSystem.cs:257]: Nope, it does damage - damage = 0, it's the same formula used on infected zombies in the same class.

Comment (Skyedra) [reply Content.Server/Zombies/ZombieSystem.cs:257]: Oh, huh, interesting.  Thanks for the explanation

**Comment by Skyedra** [comment]
Thank you for the PR.  For what it's worth, here are my thoughts on zombies:

Regarding the friendly fire portion:  I am in favor of preventing friendly fire between zombies.  It's very easy for zombies to get confused who is a zombie and who isn't, and I've accidentally been mowed down as a zombie by a mob of friendly zombies myself.  If the hit effect is prevented entirely, it will hopefully cue zombies that they're attacking a fellow zombie and they'll focus their efforts elsewhere.

Regarding the AME portion:  Short version is I agree an electric shock as suggested above (or otherwise something that stuns) would be the best route.  

Long version is that fundamentally, it's not very engaging to RP as a braindead creature.  Entertaining the first round, sure, but after that it doesn't stay engaging because the player is not supposed to make any meaningful decisions as a zombie.  Without any meaningful decisions or interactions, gameplay becomes stale and I think is a major contributing factor to why people disliked zombie rounds.

I think less-dumb zombies would be much more interesting to play as.  Limited communication and abilities, but still able to work together and pose a challenge to the station.  Zombies should be easily defeated, their principle strategies being either to quickly sneak in and convert humans, or to work together to take down larger targets like infrastructure/sec.  

To that end, a non-lethal shock would still allow zombies to work together to make decisions to go after infrastructure should they wish to go that route -- while requiring teamwork or otherwise a lot of time between stuns.  

Here are some specific design thoughts on zombies vs infrastructure:
1) If a single human armed with one of the tools found in the engi area can't reliably defend the AME against a single zombie, the zombie stats should probably be further nerfed to a point where this works out.  (Last time I played a zombie round, it was basically impossible to survive as a human without a ranged weapon.)
2) Additionally, a zombie, perhaps even a couple zombies, probably shouldn't be able to get through an electrified door easily.  It should take a large group effort / considerable time to get through these.  Perhaps knocking them out for an extended time after each bite to give humans time to notice.
3) Finally, if AME *is* taken down, perhaps it should be easier to call in for rescue (or even mechanisms to repair AME) so players aren't stuck indefinitely on a doomed station?

**Comment by John-Doom** [comment]
> Thank you for the PR. For what it's worth, here are my thoughts on zombies...

Thanks for your comment. Some of your suggestions are outside the scope of this PR, but I agree with you. I'd also add that I don't like playing as zombie, I'd prefer them to be AIs, but...

**Comment by John-Doom** [comment]
@metalgearsloth can this PR be merged? It works. The only thing I'm sorry is the fact you mentioned, that I had to hardcode AMEengine in ElectrocutionSystem.cs, but it's a temporary fix to an existing bug that I don't know how to fix myself :(

**Comment by Equivocateur** [comment]
temporary fix pls merge :pleading_face: :pray:

**Comment by EmoGarbage404** [comment]
change the name and description, considering the implementation is completely different.

**Comment by John-Doom** [comment]
> change the name and description, considering the implementation is completely different.

Name matches current implementation, what do you mean? o.O

**Comment by John-Doom** [comment]
> "webedit temporary fix pls merge 🥺 🙏 "

If you mean the AME Shields bug fix, I could exclude it from the PR and focus it just on the zombie vs zombies stuff, but there is no need to be sarcastic.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This needs to be a new component for faction damage.

**Pull request closed without merging**
